storage/innobase/log/log0log.cc

```diff
@@ -132,6 +132,7 @@ bool log_t::create()
 #endif
 
   latch.SRW_LOCK_INIT(log_latch_key);
+  lsn_lock.init();
```

Comment:
This is now corrects the removal of `init_lsn_lock()` in MDEV-33515. 

1. Looking further at the changes done by this MDEV I see it also removed `destroy_lsn_lock()` in `log_t::close` which should be done to destroy the pthread mutex object at the end. So, we should add a call to `lsn_lock.destory` in `log_t::close`.
2. I would suggest to add a debug bool variable `initialized` for both mutex and latch implementation and assert while acquiring. This would detect the problem immediately in debug run. 
3. To allow us to detect such issues in linux, I think we should have a debug way to run different mutex implementations and have some mtr for them.

Reply:
1. You are right, we’d better have also this to avoid potential resource leaks:
```diff
diff --git a/storage/innobase/log/log0log.cc b/storage/innobase/log/log0log.cc
index 6fdb8499d23..93f8db6bfc0 100644
--- a/storage/innobase/log/log0log.cc
+++ b/storage/innobase/log/log0log.cc
@@ -1331,6 +1331,7 @@ void log_t::close()
 #endif
 
   latch.destroy();
+  lsn_lock.destroy();
 
   recv_sys.close();
 
```
2. This is a good idea, but I think that it should be done separately in the 10.6 branch. This is not entirely trivial. We have a problem that the `sux_lock::readers_lock` in debug builds is being reinitialized. We would need something like this. This is not complete; at least the test `innodb.undo_truncate` would fail because `trx_sys_create_rsegs()` is reinitializing a latch in `trx_rseg_t::init()` after `srv_undo_tablespaces_reinit()` had already done it properly (without a prior `destroy()`):
```diff
diff --git a/storage/innobase/include/srw_lock.h b/storage/innobase/include/srw_lock.h
index f952722426d..fd7e08dc3be 100644
--- a/storage/innobase/include/srw_lock.h
+++ b/storage/innobase/include/srw_lock.h
@@ -30,7 +30,7 @@ this program; if not, write to the Free Software Foundation, Inc.,
 # define SUX_LOCK_GENERIC /* fall back to generic synchronization primitives */
 #endif
 
-#if !defined SUX_LOCK_GENERIC && 0 /* defined SAFE_MUTEX */
+#if !defined SUX_LOCK_GENERIC // && 0 /* defined SAFE_MUTEX */
 # define SUX_LOCK_GENERIC /* Use dummy implementation for debugging purposes */
 #endif
 
@@ -39,15 +39,26 @@ template<bool spinloop>
 class pthread_mutex_wrapper final
 {
   pthread_mutex_t lock;
+#ifdef UNIV_DEBUG
+  bool initialized{false};
 public:
-  void init()
+  ~pthread_mutex_wrapper() { ut_ad(!initialized); }
+#endif
+public:
+  void init(bool maybe_reinit= false)
   {
+    ut_ad(maybe_reinit || !initialized);
+    ut_d(initialized= true);
     if (spinloop)
       pthread_mutex_init(&lock, MY_MUTEX_INIT_FAST);
     else
       pthread_mutex_init(&lock, nullptr);
   }
-  void destroy() { pthread_mutex_destroy(&lock); }
+  void destroy()
+  {
+    ut_ad(initialized); ut_d(initialized=false);
+    pthread_mutex_destroy(&lock);
+  }
 # ifdef PTHREAD_ADAPTIVE_MUTEX_INITIALIZER_NP
   void wr_lock() { pthread_mutex_lock(&lock); }
 # else
diff --git a/storage/innobase/include/sux_lock.h b/storage/innobase/include/sux_lock.h
index a8f51fdcdba..fe2562ff0ba 100644
--- a/storage/innobase/include/sux_lock.h
+++ b/storage/innobase/include/sux_lock.h
@@ -294,7 +294,7 @@ template<> inline void sux_lock<ssux_lock_impl<true>>::init()
   lock.init();
   ut_ad(!writer.load(std::memory_order_relaxed));
   ut_ad(!recursive);
-  ut_d(readers_lock.init());
+  ut_d(readers_lock.init(true));
 #ifdef UNIV_DEBUG
   if (auto r= readers.load(std::memory_order_relaxed))
     ut_ad(r->empty());
diff --git a/storage/innobase/sync/srw_lock.cc b/storage/innobase/sync/srw_lock.cc
index 6c37d120f73..5db6e627590 100644
--- a/storage/innobase/sync/srw_lock.cc
+++ b/storage/innobase/sync/srw_lock.cc
@@ -661,6 +661,7 @@ void srw_lock_debug::destroy()
     ut_ad(r->empty());
     delete r;
   }
+  readers_lock.destroy();
   srw_lock::destroy();
 }
 
```
3. It is possible to add `-DSUX_LOCK_DEBUG` to `CMAKE_CXX_FLAGS`, or if you do not want to recompile everything, you could patch the start of `srw_lock.h` as shown above.

Reply:
I didn't mean manual run. It would be useful if we could automate it but it is not necessarily be done part of this patch.